### PR TITLE
feat: Geant4 detector plugin, part 1

### PR DESCRIPTION
**Pull request by asalzburger**
This PR does some preparation work for the purely Geant4 detectors.

It introduces:
  *  the `KDTreeTrackingGeometryBuilder` as described in the docs
  * a `ProtoDetector` description similar to the `ProtoLayer` (the `TGeoPlugin` should eventually also change to this)
  * adds some output formatting
  * changes explicit construction of `CylinderLayer` and `DiscLayer` in their respective `create(...)` method rather than the base class.

And it harmonises the evelope definitions between the duplicated `ProtoLayer::envelope` and `Extent::envelope` - they were both the same, once defined with `std::vector<std::pair<real,real>>` and once with `std::array<std::array<real, . > .>`.

### CODE REVIEW

File: Core/include/Acts/Geometry/Extent.hpp [hunk 1]
```diff
@@ -181,8 +198,9 @@ class Extent {
   bool constrains(BinningValue bValue = binValues) const;
 
   /// Convert to output stream for screen output
-  /// @param sl [in,out] The output stream
-  std::ostream& toStream(std::ostream& sl) const;
```
Comment (andiwand) [inline Core/include/Acts/Geometry/Extent.hpp:185]: is this completely obsolete or should we keep it and implement `toString` with he help of this function?

File: Core/include/Acts/Geometry/KDTreeTrackingGeometryBuilder.hpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  ///
+  /// @param [in] cfg is the configuration struct for this builder
+  /// @param [in] logger logging instance
+  KDTreeTrackingGeometryBuilder(
+      const Config& cfg,
+      std::unique_ptr<const Logger> logger =
+          getDefaultLogger("KDTreeTrackingGeometryBuilder", Logging::INFO));
+
+  /// Destructor
+  ~KDTreeTrackingGeometryBuilder() override = default;
```
Comment (andiwand) [inline Core/include/Acts/Geometry/KDTreeTrackingGeometryBuilder.hpp:65]: this one could be omitted as it should be defaulted by default

File: Core/include/Acts/Geometry/CuboidVolumeBuilder.hpp [hunk 3]
```diff
@@ -72,7 +72,7 @@ class CuboidVolumeBuilder : public ITrackingVolumeBuilder {
     // Bins in Z direction
     size_t binsZ = 1;
     // Envelope in X (along layer normal)
-    std::pair<double, double> envelopeX{0, 0};
+    std::array<ActsScalar, 2u> envelopeX{0, 0};
```
Comment (andiwand) [inline Core/include/Acts/Geometry/CuboidVolumeBuilder.hpp:75]: could this be a range 1d?

File: Core/src/Geometry/ProtoDetector.cpp [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+      if (layersPresent) {
+        break;
+      }
+    }
+
+    // If layers are present, it can't be a container in the legacy style
+    layerContainer = layersPresent;
+    auto binValue = constituentBinning[0].binvalue;
+    // Set the first last
+    auto& fVolume = *constituentVolumes.begin();
```
Comment (andiwand) [inline Core/src/Geometry/ProtoDetector.cpp:71]:
```suggestion
    auto& fVolume = constituentVolumes.front();
```

File: Core/src/Geometry/ProtoDetector.cpp [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  return ss.str();
+}
+
+std::string Acts::ProtoDetector::toString(const std::string& indent) const {
+  std::string subIndent("  ");
+  std::stringstream ss;
+  ss << indent << "> detector: " << name << '\n';
+  ss << worldVolume.toString(indent + subIndent) << '\n';
+  return ss.str();
+}
```
Comment (andiwand) [inline Core/src/Geometry/ProtoDetector.cpp:210]: nitpick: new line at the end

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  std::vector<BinningData> layerSurfaceBinning = {};
+  /// Internal structure container
+  std::vector<ProtoVolume> constituentVolumes = {};
+  /// The constituent binning if this a container
+  std::vector<BinningData> constituentBinning = {};
+
+  /// Define an operator==
+  ///
+  /// @param ptVolume the proto volume to be checked
+  bool operator==(const ProtoVolume& ptVolume) const;
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:42]: should we also add an `=!` operator?

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  void harmonize(bool legacy = true) {
+    worldVolume.extendUp(worldVolume);
+    worldVolume.constrainDown(worldVolume);
+    worldVolume.harmonize(legacy);
+  }
+
+  /// Define an operator==
+  ///
+  /// @param pd the proto detector to be checked
+  bool operator==(const ProtoDetector& pd) const {
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:97]: should we also add an `=!` operator?

File: Core/src/Geometry/ProtoLayer.cpp [hunk 8]
```diff
@@ -35,14 +35,14 @@ ProtoLayer::ProtoLayer(
 
 double ProtoLayer::min(BinningValue bval, bool addenv) const {
   if (addenv) {
-    return extent.min(bval) - envelope[bval].first;
+    return extent.min(bval) - std::abs(envelope[bval][0u]);
```
Comment (andiwand) [inline Core/src/Geometry/ProtoLayer.cpp:38]: why is `abs` necessary here?

**Review by andiwand — COMMENTED** [review]
went through once but need to take a deeper look at the kd tree and proto volume stuff

File: Core/include/Acts/Geometry/CuboidVolumeBuilder.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/CuboidVolumeBuilder.hpp:75]: Potentially, though as this is the access to the user an `{x,y}` initialisation is rather handy.

File: Core/src/Geometry/ProtoDetector.cpp (hunk 4 above)
Comment (asalzburger) [reply Core/src/Geometry/ProtoDetector.cpp:71]: Fixed.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 6 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:42]: Done.

File: Core/src/Geometry/ProtoDetector.cpp (hunk 5 above)
Comment (asalzburger) [reply Core/src/Geometry/ProtoDetector.cpp:210]: Added.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 7 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:97]: added.

File: Core/src/Geometry/ProtoLayer.cpp (hunk 8 above)
Comment (asalzburger) [reply Core/src/Geometry/ProtoLayer.cpp]: Actually, to be completely save, it should also be applied at the `max()`, but I guess, what I will do is remove this, and added potentially a check.

**Comment by asalzburger** [comment]
I think I have addressed all the comments now.

**Comment by andiwand** [comment]
I will take one more look and then we can move on to the sequel

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+#include <string>
+
+namespace Acts {
+
+/// A proto volume description being used to define an overall
+/// structure of either a TrackingVolume or Experimental::DetectorVolume
+struct ProtoVolume {
+  /// Name of the proto volume
+  std::string name = "";
+  /// The extent of this layer
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:25]:
```suggestion
  /// The extent of this volume
```
I got confused here. does the volume also have a single surface?

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  Extent extent;
+  /// Boolean to indicate this is a container, needed for legacy
+  bool layerContainer = false;
+
+  /// Set the layer type to indicate the layer volume
+  Acts::Surface::SurfaceType layerType = Acts::Surface::SurfaceType::Other;
+  /// The surface binninng fo the layer
+  std::vector<BinningData> layerSurfaceBinning = {};
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:33]: should we bundle all the legacy information into a legacy struct?

e.g.

```
struct Legacy {
  /// Boolean to indicate this is a container, needed for legacy
  bool layerContainer = false;
  /// Set the layer type to indicate the layer volume
  Acts::Surface::SurfaceType layerType = Acts::Surface::SurfaceType::Other;
  /// The surface binninng fo the layer
  std::vector<BinningData> layerSurfaceBinning = {};

  // ...
}

std::optional<Legacy> legacy;
```

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  /// if off it creates a description for `Acts::Detector`.
+  void harmonize(bool legacy = true);
+
+  /// Extend the tracking volume with the its own constituents,
+  /// upwards here means that extens are promoted to the mother
+  ///
+  /// @param ptVolume the protoVolume
+  void extendUp(ProtoVolume& ptVolume);
+
+  /// Extend the tracking volume with the its own constituents
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:57]:
```suggestion
  /// Extend the tracking volume with its own constituents
```

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  /// @param bValue the binning value that is propagated
+  void propagateMinDown(BinningValue bValue);
+
+  /// Extend the tracking volume with the its own constituents
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:61]:
```suggestion
  /// Extend the tracking volume with its own constituents
```

File: Core/include/Acts/Geometry/ProtoDetector.hpp [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+  /// Harmonize the detector information, this can run in two
+  /// modes, steered by the @param legacy boolean
+  ///
+  /// The legacy mode prepares everything for `Acts::TrackingVolume`,
+  /// if off it creates a description for `Acts::Detector`.
+  void harmonize(bool legacy = true);
+
+  /// Extend the tracking volume with the its own constituents,
+  /// upwards here means that extens are promoted to the mother
```
Comment (andiwand) [inline Core/include/Acts/Geometry/ProtoDetector.hpp:52]:
```suggestion
  /// Extend the tracking volume with its own constituents,
  /// upwards here means that extents are promoted to the mother
```

File: Core/src/Geometry/KDTreeTrackingGeometryBuilder.cpp [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+                                                   bType1, pLayer);
+    }
+  }
+  if (tLayer != nullptr and tLayer->representingVolume() != nullptr) {
+    ACTS_DEBUG(indent << "> translated into layer bounds: "
+                      << tLayer->representingVolume()->volumeBounds());
+  }
+
+  return tLayer;
+}
```
Comment (andiwand) [inline Core/src/Geometry/KDTreeTrackingGeometryBuilder.cpp:217]: nitpick: new line at the end

File: Core/src/Geometry/ProtoDetector.cpp [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+  // Deal with the constituents
+  if (not constituentVolumes.empty()) {
+    if (constituentBinning.empty()) {
+      std::string errorMsg = std::string("ProtoVolume '") + name +
+                             std::string("' with constituents, but no binning");
+      throw std::runtime_error(errorMsg);
+    }
+
+    // For legacy volumes, check if layers are present
+    bool layersPresent = layerContainer;
```
Comment (andiwand) [inline Core/src/Geometry/ProtoDetector.cpp:58]: couldn't we use `layerContainer` directly and ditch `layersPresent`?

File: Core/src/Geometry/ProtoDetector.cpp [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    std::vector<float> borders = {};
+
+    // The volumes should be harmonized in all other constraining values
+    for (auto obValue : s_binningValues) {
+      if (obValue != binValue and extent.constrains(obValue)) {
+        otherConstrains.push_back(obValue);
+      }
+    }
+
+    // Legacy  conversion - layers are kept untouched
```
Comment (andiwand) [inline Core/src/Geometry/ProtoDetector.cpp:83]:
```suggestion
    // Legacy conversion - layers are kept untouched
```

**Review by andiwand — APPROVED** [review]

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 10 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:33]: Oh, that's a good suggestion.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 9 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:25]: Fixed.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 10 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:33]: Actually, it only needs a `legacyLayerType`, the other information is either both legacy or new, or redundant.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 11 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:57]: Fixed.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 13 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:52]: Fixed.

File: Core/src/Geometry/KDTreeTrackingGeometryBuilder.cpp (hunk 14 above)
Comment (asalzburger) [reply Core/src/Geometry/KDTreeTrackingGeometryBuilder.cpp:217]: Inserted, probably we can make `clang-format` to check and fix this?

File: Core/src/Geometry/ProtoDetector.cpp (hunk 15 above)
Comment (asalzburger) [reply Core/src/Geometry/ProtoDetector.cpp:58]: Not really, as `layerPresent` is is an `XAND` of all the possible layer containers, anyhow it is a bit changed now.

File: Core/src/Geometry/ProtoDetector.cpp (hunk 16 above)
Comment (asalzburger) [reply Core/src/Geometry/ProtoDetector.cpp:83]: Fixed.

File: Core/include/Acts/Geometry/ProtoDetector.hpp (hunk 12 above)
Comment (asalzburger) [reply Core/include/Acts/Geometry/ProtoDetector.hpp:61]: Fixed.

**Comment by asalzburger** [comment]
I requested re-approval, since I changed the `ProtoDetector` following/initiated by your suggestions.

**Comment by asalzburger** [comment]
@andiwand - can you have a second look over it, please?

**Comment by andiwand** [comment]
having a look right now @asalzburger

**Review by andiwand — APPROVED** [review]
lets get this in and move on to part two :tada:

**Pull request merged**